Vault token renew time

### Nomad version

Nomad v1.11.1
BuildDate 2025-12-09T20:10:56Z
Revision 5b76eb0535615e32faf4daee479f7155ea16ec0d

### Operating system and environment details

Ubuntu 24.04.1 LTS

### Nomad configuration

We have a cluster consisting of 3 servers and several clients. The cluster is integrated with Vault.

```hcl
vault {
  enabled = true
  address = "<vault_addr>"

  default_identity {
    aud = ["vault.io"]
    ttl = "1h"
  }
}
```

### Vault role configuration

```text
allowed_redirect_uris      <nil>
bound_audiences            [vault.io]
bound_claims               map[]
bound_claims_type          string
bound_subject              n/a
claim_mappings             map[nomad_job_id:nomad_job_id nomad_namespace:nomad_namespace nomad_task:nomad_task]
clock_skew_leeway          0
expiration_leeway          0
groups_claim               n/a
max_age                    0
not_before_leeway          0
oidc_scopes                <nil>
role_type                  jwt
token_bound_cidrs          []
token_explicit_max_ttl     0s
token_max_ttl              0s
token_no_default_policy    false
token_num_uses             0
token_period               30m
token_policies             [<policies_list>]
token_ttl                  0s
token_type                 service
user_claim                 /nomad_job_id
user_claim_json_pointer    true
verbose_oidc_logging       false
```

### Issue

We experienced a short network outage and connectivity issues between the Nomad cluster and the Vault cluster.

Initially, we observed warnings indicating that Nomad could not read secrets from Vault:

```text
2026-05-06T05:26:44.736Z [WARN]  agent: (view) vault.read(kv2/common/apps/loki/basic_auth_credentials): vault.read(kv2/common/apps/loki/basic_auth_credentials): context deadline exceeded (retry attempt 2 after "500ms")

2026-05-06T05:23:39.046Z [WARN]  agent: (view) vault.read(kv2/common/apps/loki/basic_auth_credentials): vault.read(kv2/common/apps/loki/basic_auth_credentials): context deadline exceeded (retry attempt 1 after "250ms")
```

Then, approximately 15 minutes later, we started seeing Vault token renewal failures:

```text
2026-05-06T05:40:53.562Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=4e5fc036-f585-2107-a06e-8d105cffc05d task=vector type="Restart Signaled" msg="Vault: new Vault token acquired" failed=false

2026-05-06T05:40:53.406Z [ERROR] client.alloc_runner.task_runner.task_hook.vault: failed to renew Vault token: alloc_id=4e5fc036-f585-2107-a06e-8d105cffc05d task=vector

error=
| failed to renew the vault token: Error making API request.
| URL: PUT https://<vault_server>/v1/auth/token/renew-self
| Code: 403. Errors:
| * 2 errors occurred:
|   * permission denied
|   * invalid token
```

After this, the allocation was restarted with the reason:

```text
Vault: new Vault token acquired
```

However, no new secret was actually acquired. The allocation appears to be restarted only because the Vault token renewal failed and the original token expired.

Our assumption is that because `token_period` is set to `30m`, Nomad attempts token renewal only once at half of the lease duration (15 minutes in our case).

We found this logic in the source code:

https://github.com/hashicorp/nomad/blob/v1.11.1/client/vaultclient/vaultclient.go#L702

As a result, if Nomad experiences temporary connectivity issues with Vault during that renewal attempt, it does not appear to retry renewal later before the token expires.

Eventually, after the token period expires, Vault invalidates the token and Nomad receives a `403 invalid token` response. Nomad then restarts the allocation with the message `"Vault: new Vault token acquired"`.

This behavior seems problematic because temporary Vault connectivity issues during a single renewal window can cause unnecessary allocation restarts.

Is this expected behavior, or is there a way to configure Nomad to retry Vault token renewal more frequently before token expiration?
If this is expected behavior, it seems like Nomad should retry token renewal more aggressively before allowing the token to expire.

@mismithhisler Yeah, there were such lines in the logs as well right before the 403:

```text
2026-05-05T10:24:42.985Z [ERROR] client.vault: error during renewal of lease or token failed due to a non-fatal error; retrying: name=default error="failed to renew the vault token: context deadline exceeded" period="2026-05-05 10:39:46.985064248 +0000 UTC m=+14465.801188360"

--- a/.changelog/27947.txt
+++ b/.changelog/27947.txt
@@ -1,0 +1,3 @@
+```release-note:improvement
+vault: adds token renewal retries
+```

--- a/client/vaultclient/vaultclient.go
+++ b/client/vaultclient/vaultclient.go
@@ -409,6 +409,12 @@ func (c *vaultClient) renew(req *vaultClientRenewalRequest) error {
 			strings.Contains(errMsg, "token not found") {
 			fatal = true
 		} else {
+			// In the event of a non fatal error, retry in a minute.
+			// TODO: mismithhisler - This is a temporary fix until a subsequant
+			// refactor of the vault client is merged that contains proper
+			// exponential backoffs.
+			next = time.Now().Add(time.Minute)
+
 			c.logger.Debug("renewal error details", "req.increment", req.increment, "lease_duration", leaseDuration, "renewal_duration", renewalDuration)
 			c.logger.Error("error during renewal of lease or token failed due to a non-fatal error; retrying",
 				"error", renewalErr, "period", next)

--- a/client/vaultclient/vaultclient.go
+++ b/client/vaultclient/vaultclient.go
@@ -410,7 +410,7 @@ func (c *vaultClient) renew(req *vaultClientRenewalRequest) error {
 			fatal = true
 		} else {
 			// In the event of a non fatal error, retry in a minute.
-			// TODO: mismithhisler - This is a temporary fix until a subsequant
+			// TODO: mismithhisler - This is a temporary fix until a followup
 			// refactor of the vault client is merged that contains proper
 			// exponential backoffs.
 			next = time.Now().Add(time.Minute)

--- a/client/vaultclient/vaultclient.go
+++ b/client/vaultclient/vaultclient.go
@@ -1,4 +1,4 @@
-// Copyright IBM Corp. 2015, 2025
+// Copyright IBM Corp. 2015, 2026
 // SPDX-License-Identifier: BUSL-1.1
 
 package vaultclient

--- a/client/vaultclient/vaultclient.go
+++ b/client/vaultclient/vaultclient.go
@@ -413,7 +413,7 @@ func (c *vaultClient) renew(req *vaultClientRenewalRequest) error {
 			// TODO: mismithhisler - This is a temporary fix until a followup
 			// refactor of the vault client is merged that contains proper
 			// exponential backoffs.
-			next = time.Now().Add(time.Minute)
+			next = min(time.Now().Add(time.Minute), next)
 
 			c.logger.Debug("renewal error details", "req.increment", req.increment, "lease_duration", leaseDuration, "renewal_duration", renewalDuration)
 			c.logger.Error("error during renewal of lease or token failed due to a non-fatal error; retrying",

--- a/client/vaultclient/vaultclient.go
+++ b/client/vaultclient/vaultclient.go
@@ -409,11 +409,14 @@ func (c *vaultClient) renew(req *vaultClientRenewalRequest) error {
 			strings.Contains(errMsg, "token not found") {
 			fatal = true
 		} else {
-			// In the event of a non fatal error, retry in a minute.
+			// In the event of a non fatal error, retry in a maximum of 1 minute
 			// TODO: mismithhisler - This is a temporary fix until a followup
 			// refactor of the vault client is merged that contains proper
 			// exponential backoffs.
-			next = min(time.Now().Add(time.Minute), next)
+			retry := time.Now().Add(time.Minute)
+			if retry.Before(next) {
+				next = retry
+			}
 
 			c.logger.Debug("renewal error details", "req.increment", req.increment, "lease_duration", leaseDuration, "renewal_duration", renewalDuration)
 			c.logger.Error("error during renewal of lease or token failed due to a non-fatal error; retrying",

